File: BEmuJava/bemu/src/main/java/com/bloomberglp/blpapi/HistoricElementFieldExceptions.java
//------------------------------------------------------------------------------
// <copyright project="BEmu_maven" file="/BEmu_maven/bemu/src/main/java/com/bloomberglp/blpapi/HistoricElementFieldExceptions.java" company="Jordan Robinson">
//     Copyright (c) 2013 Jordan Robinson. All rights reserved.
//
//     The use of this software is governed by the Microsoft Public License
//     which is included with this distribution.
// </copyright>
//------------------------------------------------------------------------------

package com.bloomberglp.blpapi;

import com.bloomberglp.blpapi.Element;
import com.bloomberglp.blpapi.Name;
import com.bloomberglp.blpapi.SchemaTypeDefinition;

public class HistoricElementFieldExceptions extends Element
{
    private final HistoricElementString _fieldId;
    private final HistoricElementErrorInfo _errorInfo;

    public HistoricElementFieldExceptions(String badField)
    {
        this._fieldId = new HistoricElementString("fieldId", badField);
        this._errorInfo = new HistoricElementErrorInfo();
    }
    
    public SchemaTypeDefinition typeDefinition() throws Exception
    {
    	return new SchemaTypeDefinition(this.datatype(), new Name("FieldException"));
    }
	
    public Name name() throws Exception
    {
    	return new Name("errorInfo");
    }
    
    public int numValues()
    {
    	return 1;
    }
    
    public int numElements()
    {
    	return 2;
    }
    
    public String getElementAsString(String name) throws Exception
    {
    	return this.getElement(name).getValueAsString();
    }
    
    public int getElementAsInt32(String name) throws Exception
    {
    	return this.getElement(name).getValueAsInt32();
    }
    
    public Element getElement(String name) throws Exception
    {
    	if(name.equals("fieldId"))
    		return this._fieldId;
    	
    	else if(name.equals("errorInfo"))
    		return this._errorInfo;
    	
    	else
    		return super.getElement(name);
    }
    
    public boolean hasElement(String name)
    {
    	return name.equals("fieldId") || name.equals("errorInfo");
    }

    protected StringBuilder prettyPrint(int tabIndent) throws Exception
    {
        String tabs = com.bloomberglp.blpapi.IndentType.Indent(tabIndent);
        StringBuilder result = new StringBuilder();
        
        result.append(String.format("%s%s = {%s", tabs, this.name().toString(), System.getProperty("line.separator")));
        result.append(this._fieldId.prettyPrint(tabIndent + 1));
        result.append(this._errorInfo.prettyPrint(tabIndent + 1));
        result.append(String.format("%s}%s", tabs, System.getProperty("line.separator")));

        return result;
    }
    
    public boolean isComplexType()
    {
    	return true;
    }
    
    public boolean isArray()
    {
    	return false;
    }
    
    public boolean isNull()
    {
    	return false;
    }
    
    
}

File: BEmu_cpp/src/HistoricalDataRequest/HistoricElementFieldExceptions.cpp
//------------------------------------------------------------------------------
// <copyright project="BEmu_cpp" file="src/HistoricalDataRequest/HistoricElementFieldExceptions.cpp" company="Jordan Robinson">
//     Copyright (c) 2013 Jordan Robinson. All rights reserved.
//
//     The use of this software is governed by the Microsoft Public License
//     which is included with this distribution.
// </copyright>
//------------------------------------------------------------------------------

#include "HistoricalDataRequest/HistoricElementFieldExceptions.h"
#include "HistoricalDataRequest/HistoricElementString.h"
#include "HistoricalDataRequest/HistoricElementErrorInfo.h"

#include "BloombergTypes/Name.h"
#include "Types/IndentType.h"

namespace BEmu
{
	namespace HistoricalDataRequest
	{
		HistoricElementFieldExceptions::HistoricElementFieldExceptions(const std::string& badField) :
			_fieldId(new HistoricElementString("fieldId", badField)),
			_errorInfo(new HistoricElementErrorInfo())
		{
		}

		HistoricElementFieldExceptions::~HistoricElementFieldExceptions()
		{
		}


		Name HistoricElementFieldExceptions::name() const
		{
			Name result("fieldExceptions");
			return result;
		}

		size_t HistoricElementFieldExceptions::numValues() const
		{
			return 1;
		}

		size_t HistoricElementFieldExceptions::numElements() const
		{
			return 2;
		}

		SchemaElementDefinition HistoricElementFieldExceptions::elementDefinition() const
		{
			::blpapi_DataType_t dtype = (::blpapi_DataType_t)this->datatype();
			SchemaElementDefinition result(dtype, Name("FieldException"));
			return result;
		}

		
		bool HistoricElementFieldExceptions::isNull() const
		{
			return false;
		}

		bool HistoricElementFieldExceptions::isArray() const
		{
			return false;
		}

		bool HistoricElementFieldExceptions::isComplexType() const
		{
			return true;
		}


		boost::shared_ptr<ElementPtr> HistoricElementFieldExceptions::getElement(const char* name) const
		{
			if(strncmp(name, "errorInfo", 10) == 0)
				return boost::dynamic_pointer_cast<ElementPtr>(this->_errorInfo);

			else if(strncmp(name, "fieldId", 8) == 0)
				return boost::dynamic_pointer_cast<ElementPtr>(this->_fieldId);

			else
				throw elementPtrEx;
		}

		bool HistoricElementFieldExceptions::hasElement(const char* name, bool excludeNullElements) const
		{
			return
				(strncmp(name, "errorInfo", 10) == 0) ||
				(strncmp(name, "fieldId", 8) == 0);
		}

		int HistoricElementFieldExceptions::getElementAsInt32(const char* name) const
		{
			return this->getElement(name)->getValueAsInt32(0);
		}

		const char* HistoricElementFieldExceptions::getElementAsString(const char* name) const
		{
			return this->getElement(name)->getValueAsString(0);
		}


		std::ostream& HistoricElementFieldExceptions::print(std::ostream& stream, int level, int spacesPerLevel) const
		{
			std::string tabs(IndentType::Indent(level, spacesPerLevel));

			stream << tabs << "fieldExceptions = {" << std::endl;
			
			this->_fieldId->print(stream, level + 1, spacesPerLevel);
			this->_errorInfo->print(stream, level + 1, spacesPerLevel);

			stream << tabs << '}' << std::endl;

			return stream;
		}

	}
}